in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+|                      |                                                                                               |
+| -------------------- | --------------------------------------------------------------------------------------------- |
+| Issue                | [Refactor Contract Interactions](https://github.com/AztecProtocol/aztec-packages/issues/6973) |
+| Owners               | @just-mitch                                                                                   |
+| Approvers            | @PhilWindle @LHerskind                                                                        |
+| Target Approval Date | 2024-06-21                                                                                    |
+
+
+## Executive Summary
+
+This is a refactor of the API for interacting with contracts to improve the user (developer) experience, focused on aztec.js.
+
+The refactored approach mimics Viem's API, with some enhancements and modifications to fit our needs.
+
+In a nutshell, by being more verbose in the API, we can remove a lot of complexity and make the code easier to understand and maintain; this also affords greater understanding and control over the lifecycle of contracts and transactions.
+
+Key changes:
+- the wallet is the central point of interaction to simulate/prove/send transactions instead of `BaseContractInteraction`
+- unified api for deploying contracts, deploying accounts, and calling functions
+- idempotent simulations
+- enables more sophisticated gas estimation mechanism
+
+### Major Interface Update
+
+Remove `BaseContractInteraction` and its subclasses.
+
+Define the following:
+
+```ts
+export interface UserAPI {
+  getTxExecutionRequest(userRequest: UserRequest): Promise<TxExecutionRequest>;
+  simulate(userRequest: UserRequest): Promise<SimulationOutput>;
+  read(userRequest: UserRequest): Promise<SimulationOutput>;
+  prove(userRequest: UserRequest): Promise<UserRequest>;
+  send(userRequest: UserRequest): Promise<SentTx>;
+}
+
+export type Wallet = UserAPI & AccountInterface & PXE & AccountKeyRotationInterface;
+```
+
+### Account Creation
+
+```ts
+const pxe = await createCompatibleClient(rpcUrl, debugLogger);
+const secretKey = Fr.random();
+const salt = 42;
+
+// Returns the wallet and deployment options. 
+// Does not modify PXE state.
+const { wallet: aliceWallet, deploymentArgs } = await getSchnorrWallet(pxe, secretKey, salt);
+
+const aliceContractInstance = getContractInstanceFromDeployParams(
+  SchnorrAccountContract.artifact, 
+  deploymentArgs
+);
+
+// Register the account in PXE, and wait for it to sync.
+await aliceWallet.registerAccountInPXE({ sync: true });
```

Comment:
True. This would probably be a two step:
```ts
aliceWallet.registerAccountInPXE()
// later, in some high level context in the UI
aliceWallet.waitForAccountSynch()
// show loading spinner
```